base_layer/core/src/validation/pipeline.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright 2019. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use super::Validation;
+use crate::validation::error::ValidationError;
+use tari_transactions::transaction::Transaction;
+use crate::blocks::Block;
+
+type Pipelined = Box<dyn Validation>;
+
+/// A validation pipeline. Multiple validators (structs that implement [Validation]) can be chained together to form a
+/// validation pipeline. `ValidationPipeline implements [Validation] itself, and the pipeline is valid _if and only
+/// if_ every validator it contains is valid.
+///
+/// # Example
+///
+/// ```edition2018
+///     use tari_core::validation::{mocks::MockValidator, Validation, ValidationPipeline};
+///
+///     let mut pipeline = ValidationPipeline::new();
+///     pipeline.push(Box::new(MockValidator::new(true)));
+///     assert!(pipeline.validate().is_ok());
+/// ```
+
+pub struct ValidationPipeline {
+    validators: Vec<Pipelined>,
+}
+
+impl Default for ValidationPipeline {
+    fn default() -> Self {
+        ValidationPipeline { validators: Vec::new() }
+    }
+}
+
+impl ValidationPipeline {
+    pub fn new() -> Self {
+        ValidationPipeline::default()
+    }
+
+    pub fn push(&mut self, validator: Pipelined) {
+        self.validators.push(validator)
+    }
+}
+
+impl Validation for ValidationPipeline {
+
+
+    fn validate(&mut self) -> Result<(), ValidationError> {
+        for v in &mut self.validators {
+            if let Err(e) = v.validate() {
+                return Err(e);
+            }
```

Comment:
This if let can be replaced with `v.validate()?;`, a similar approach can be used for the `validate_block` and `validate_transaction` if let statements.

Reply:
Yup, thanks.